### PR TITLE
Slash commands

### LINKED ISSUE #65: Use slash commands instead of custom command prefix
Discord recently added new API for [Slash Commands](https://discord.com/developers/docs/interactions/slash-commands), which allows Discord to suggest the command (which is way better than remembering a ton of commands (which is hard)
image
and it is way more user-friendly); create global commands or commands specific to one server, supposedly even make commands without bot on the server

### LINKED ISSUE #85: Bot does not respond to wrongly typed commands
The bot should always respond to a command if it is wrong.
 - does not know the command
 - command has the wrong number of arguments
 - some arguments are incorrect (eg. needs name of role or name of channel).

### LINKED ISSUE #198: Commands in threads
Commands don't work in threads even if the bot is there.

Not sure if we need to handle this, considering #65 slash commands will be implemented soon.

### LINKED ISSUE #352: Bot claims he doesn't know a command then shows it as an command
Bot says, he doesn't know command `::member `

image

### PR DESCRIPTION
**Pull request by stepech**
Implement DSharp's Slash commands

This implementation **requires DSharp 4.3.0** nightly. SlashCommands V2 were recently released and proper support is once again included only in nightlies - for now.

THIS PR REMOVES OLD COMMANDS AND CODE ASSOCIATED WITH THEM

Even though github shows command files as edited, they were all rewritten from scratch.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ version: "3"
 services:
   postgres-botner:
     container_name: postgres-botner
-    image: postgres:12
+    image: postgres
```
Comment (ostorc) [inline docker-compose.yml:5]: Why this change?

File: src/HonzaBotner.Discord.Services/Commands/EmoteCommands.cs [hunk 2]
```diff
@@ -1,67 +1,49 @@
...
+    [SlashCommandPermissions(Permissions.ManageEmojis)]
+    public async Task EmoteStatsCommandAsync(
+        InteractionContext ctx,
+        [Choice("perDay", 0)]
+        [Choice("total", 1)]
+        [Option("display", "Show displayed per day or total? Defaults perDay")] long showTotal = 0,
+        [Choice("all", 0)]
+        [Choice("animated", 1)]
+        [Choice("still", 2)]
+        [Option("Type", "What type of emojis to show? Defaults all")] long type = 0)
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/EmoteCommands.cs]: Could this be an enum?

File: src/HonzaBotner.Discord.Services/Commands/PinCommands.cs [hunk 3]
```diff
@@ -1,220 +1,146 @@
...
 
-            if (result.TimedOut) return;
+    [SlashCommand("delete-pins", "Unpins messages pinned with temporary pins")]
+    [SlashCommandPermissions(Permissions.ManageChannels)]
+    public async Task DeleteAllAsync(
+        InteractionContext ctx,
+        [Option("everywhere", "Do it everywhere or just in this channel? Default: false")]
+        bool everywhere = false)
+    {
+        string customId = "unpin" + ctx.User.Id + ctx.Channel.Id;
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/PinCommands.cs]:
```suggestion
        string customId = $"unpin{ctx.User.Id}{ctx.Channel.Id}";
```

File: src/HonzaBotner.Discord/ConfigGuildProvider.cs [hunk 4]
```diff
@@ -8,23 +8,21 @@ namespace HonzaBotner.Discord;
 
 public class ConfigGuildProvider : IGuildProvider
 {
-    private readonly IOptions<DiscordConfig> _config;
+    public ulong GuildId { get; }
```
Comment (ostorc) [inline src/HonzaBotner.Discord/ConfigGuildProvider.cs:11]: Not necessary change, but also not wrong :)

File: src/HonzaBotner.Discord/DiscordBot.cs [hunk 5]
```diff
@@ -95,98 +96,29 @@ private async Task Client_ClientError(DiscordClient sender, ClientErrorEventArgs
         e.Handled = true;
     }
 
-    private Task Commands_CommandExecuted(CommandsNextExtension sender, CommandExecutionEventArgs e)
+    private Task SCommands_SlashCommandExecuted(SlashCommandsExtension e, SlashCommandExecutedEventArgs args)
```
Comment (ostorc) [inline src/HonzaBotner.Discord/DiscordBot.cs]: Maybe try to come with better name for the method :D

**Review by ostorc — COMMENTED** [review]
The first phase of the review seems fine at first look. However, I am concerned that you have removed a lot of lines in the commands, can you explain why it is so?

File: docker-compose.yml (hunk 1 above)
Comment (stepech) [reply docker-compose.yml:5]: Was trying something. Do you want me to revert it back? It downloads v14 this way...

File: src/HonzaBotner.Discord/ConfigGuildProvider.cs [hunk 6]
```diff
@@ -8,23 +8,21 @@ namespace HonzaBotner.Discord;
...
 public class ConfigGuildProvider : IGuildProvider
 {
-    private readonly IOptions<DiscordConfig> _config;
+    public ulong GuildId { get; }
     private readonly DiscordClient _client;
 
     public ConfigGuildProvider(DiscordWrapper wrapper, IOptions<DiscordConfig> config)
     {
-        _config = config;
+        if (config.Value.GuildId == null)
```
Comment (stepech) [inline src/HonzaBotner.Discord/ConfigGuildProvider.cs]: is != "=="

File: src/HonzaBotner.Discord.Services/WIPCommands/FunCommands.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+            .Where(option => option != "").ToArray();
+        if (answers.Length == 0)
+        {
+            await ctx.CreateResponseAsync("Nope");
+            return;
+        }
+        Random random = new();
+        var text = new StringBuilder("I picked: ");
+        int winNumber = random.Next(answers.Length);
+        string winner = answers[winNumber];
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/WIPCommands/FunCommands.cs]: We are mixing several method of variable declarations. Pick one and stick with it.

File: src/HonzaBotner.Discord.Services/WIPCommands/ModerationCommands.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+            .WithTitle($"New Warning for {ctx.TargetMember.DisplayName}")
+            .WithCustomId(modalId)
+            .AddComponents(new TextInputComponent("Reason:", reasonId, required: true))
+            .AsEphemeral(true);
+        await ctx.CreateResponseAsync(InteractionResponseType.Modal, response);
+
+        int numberOfWarnings = await _warningService.GetNumberOfWarnings(ctx.TargetUser.Id);
+
+        var interactivity = ctx.Client.GetInteractivity();
+        var modalReason = await interactivity.WaitForModalAsync(modalId, TimeSpan.FromMinutes(10));
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/WIPCommands/ModerationCommands.cs]: Again mixing several methods of declarations.

File: src/HonzaBotner.Discord/ConfigGuildProvider.cs (hunk 6 above)
Comment (ostorc) [reply src/HonzaBotner.Discord/ConfigGuildProvider.cs]: ?

**Review by ostorc — DISMISSED** [review]
Lets prayge that it is OK, I will belive you that you have tested it and we can merge it.

**Comment by stepech** [comment]
@ostorc this is ready now for a final review and merge

File: src/HonzaBotner.Discord.Services/Commands/EmoteCommands.cs [hunk 9]
```diff
@@ -1,70 +1,54 @@
...
     }
 
-    private async Task Display<TKey>(CommandContext ctx, bool total, string[] parameters,
-        Func<CountedEmoji, TKey> comparer)
+    [SlashCommand("emotes", "Display statistics about emotes on server")]
+    [SlashCommandPermissions(Permissions.ManageEmojis)]
+    public async Task EmoteStatsCommandAsync(
+        InteractionContext ctx,
+        [Option("display", "Display as total instead of perDay?")] bool total = true,
+        [Option("Type", "What type of emojis to show? Defaults all")] DisplayTypes type = DisplayTypes.All)
```
Comment (tenhobi) [inline src/HonzaBotner.Discord.Services/Commands/EmoteCommands.cs]: Should the first parameter be unified, so it either starts with uppercase or lowercase?

File: src/HonzaBotner.Discord.Services/Commands/FunCommands.cs [hunk 10]
```diff
@@ -1,58 +1,55 @@
...
             }
+
+            text.Remove(text.Length - 2, 2);
         }
-        catch (Exception e)
+        else
         {
-            _logger.LogWarning(e, "Couldn't send a message");
-            await ctx.Message.CreateReactionAsync(DiscordEmoji.FromName(ctx.Client, ":-1:"));
+            text.Append("\nAre you bipolar or why was this necessary?!");
```
Comment (tenhobi) [inline src/HonzaBotner.Discord.Services/Commands/FunCommands.cs]: ?

**Review by tenhobi — DISMISSED** [review]
LGTM

File: src/HonzaBotner.Discord.Services/Commands/FunCommands.cs (hunk 10 above)
Comment (stepech) [reply src/HonzaBotner.Discord.Services/Commands/FunCommands.cs]: Will add more random text in the future. Idea is to question someone's ability to think clearly when he offers only one thing to choose from. Might be funny. Might not if Kalvoda uses it....

File: src/HonzaBotner.Discord.Services/Commands/EmoteCommands.cs (hunk 9 above)
Comment (stepech) [reply src/HonzaBotner.Discord.Services/Commands/EmoteCommands.cs]: Will fix

Comment (stepech) [reply src/HonzaBotner.Discord.Services/Commands/EmoteCommands.cs:42]: <@427441425861443595>

File: Packages.props [hunk 11]
```diff
@@ -2,9 +2,9 @@
     <ItemGroup>
         <PackageReference Update="Chronic.Core" Version="0.4.0" />
         <PackageReference Update="Cronos" Version="0.7.1" />
-        <PackageReference Update="DSharpPlus" Version="4.2.0" />
-        <PackageReference Update="DSharpPlus.CommandsNext" Version="4.2.0" />
-        <PackageReference Update="DSharpPlus.Interactivity" Version="4.2.0" />
+        <PackageReference Update="DSharpPlus" Version="4.3.0-nightly-01169" />
+        <PackageReference Update="DSharpPlus.Interactivity" Version="4.3.0-nightly-01169" />
```
Comment (ostorc) [inline Packages.props:6]: Is this up to date?

**Review by tenhobi — DISMISSED** [review]
LGTM

File: src/HonzaBotner.Discord.Services/Commands/BotCommands.cs [hunk 12]
```diff
@@ -1,131 +1,78 @@
...
 using Microsoft.Extensions.Options;
 
 namespace HonzaBotner.Discord.Services.Commands;
 
-[Group("bot")]
-[Description("Commands to get you info about Discord bot and other things.")]
-public class BotCommands : BaseCommandModule
+public class BotCommands : ApplicationCommandModule
 {
-    private readonly ILogger<BotCommands> _logger;
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/BotCommands.cs:18]: I am not against this change, but please be sure to have at least some logging in the commands so we can fine-tune it.

File: src/HonzaBotner.Discord/DiscordBot.cs [hunk 13]
```diff
@@ -220,6 +154,12 @@ private Task Client_ChannelCreated(DiscordClient client, ChannelCreateEventArgs
         return _eventHandler.Handle(args);
     }
 
+    private Task Client_ThreadCreated(DiscordClient client, ThreadCreateEventArgs args)
+    {
+        Task.Run(() => args.Thread.SendMessageAsync("Ping <@132599706747535360>"));
```
Comment (ostorc) [inline src/HonzaBotner.Discord/DiscordBot.cs:161]: What is this is?

File: src/HonzaBotner/appsettings.CvutFit.json [hunk 14]
```diff
@@ -11,11 +11,6 @@
     },
     "Discord": {
         "GuildId": 366970031445377024,
-        "Prefixes": [
```
Comment (ostorc) [inline src/HonzaBotner/appsettings.CvutFit.json:14]: 😢

**Review by ostorc — COMMENTED** [review]
Have you fixed the DI stuff (scoped lifetime)?

File: src/HonzaBotner.Discord/DiscordBot.cs (hunk 13 above)
Comment (tenhobi) [reply src/HonzaBotner.Discord/DiscordBot.cs:161]: lol, didnt notice :D

File: Packages.props (hunk 11 above)
Comment (stepech) [reply Packages.props:6]: yes, after merge, DependaBot is going to take care of it. RN, this should be latest.

File: src/HonzaBotner.Discord.Services/Commands/BotCommands.cs (hunk 12 above)
Comment (stepech) [reply src/HonzaBotner.Discord.Services/Commands/BotCommands.cs:18]: True, RN logging is implemented only at the top level, posting errors to discord. I will implement more logging where needed

**Review by ostorc — APPROVED** [review]
